sql-plugin/src/main/scala/org/apache/spark/sql/rapids/AggregateFunctions.scala

```diff
@@ -70,13 +70,28 @@ trait GpuAggregateFunction extends GpuExpression
   // them to initialize the aggregation buffer, and returns them in case
   // of an empty aggregate when there are no expressions,
   // here we copy them but with the gpu equivalent
-  val initialValues: Seq[GpuExpression]
+  val initialValues: Seq[Expression]
 
   // update: first half of the aggregation (count = count)
   val updateExpressions: Seq[Expression]
 
+  // expression to use to modify pre and post a cuDF update aggregate
+  // preUpdate: modify an incoming batch to match the shape/type cuDF expects
+  // postUpdate and postUpdateAttr: take the output of a cuDF update aggregate and return
+  //   what spark expects
+  lazy val preUpdate: Seq[Expression] = aggBufferAttributes
+  lazy val postUpdate: Seq[Expression] = aggBufferAttributes
+  lazy val postUpdateAttr: Seq[AttributeReference] = aggBufferAttributes
```

Comment:
I think it would be cleaner if we make `preUdate` and `postUpdate` be `NamedExpression`s. Then it tells us exactly what the attributes are that are coming out of it, and we don't need a separate postUpdateAttr.

Reply:
I am not sure I am able to do this. I tried doing this locally, but I end up using things that are not `NamedExpression`, for instance `GpuCast`. Is there way to obtain a `NamedExpression` in those scenarios?

Reply:
You wrap it with a GpuAlias. That can give a name to anything, but you have to be sure that the expression ids match, because that is what binding uses to find the same expression, not the name

Reply:
Ok, I'll give this another try with GpuAlias.

Reply:
@revans2 and I discussed on the side and we are moving towards a positional way of dealing with most of the aggregates. The main issue is it's not clear from the code what an expression should bind to. Currently we bind to attributes that mean different things, depending on the stage. 

A clearer approach is to bind when it's perfectly clear what the bind is for and in other cases assume order of the slots for the aggregation (and change the interface to take indices instead) when it is appropriate. I am working on this change, but to keep this PR smaller, I'd like to merge it as is, and continue separately with this (https://github.com/NVIDIA/spark-rapids/issues/3194)